posit-dev/positron
# Add actions for executing code

### Intent

Make it possible to create keybindings that run code in Positron. This feature has been requested frequently by both private beta and internal beta users.

Addresses https://github.com/posit-dev/positron/issues/2375
Addresses https://github.com/posit-dev/positron/issues/2661

### Approach

There are two new commands here; they have arguments and semantics that are different enough to warrant the bifurcation. 

`workbench.action.executeCode.console` - executes code as if the user typed it. This is useful for keybindings that are used  more like macros.

`workbench.action.executeCode.silently` - executes code silently. This is useful for code that is run for its side effects (e.g. making API calls). 

Both commands accept an argument object that can specify `langId` and `code` fields. They can also accept just a plain string as an argument, in which case the code will be run in whatever console happens to be active. 

### QA Notes

Here is an example keybinding to start playing with this:

```json
    {
        "key": "Alt+Shift+E",
        "command": "workbench.action.executeCode.console",
        "description": "Print Elbereth",
        "when": "editorTextFocus",
        "args": {
            "langId": "python", 
            "code": "print('Elbereth')",
            "focus": true
        }
    }
````

Try:
- executing silently and observing side effects (note that many side effects are not visible until you run something else in the console since that's when most kernels check for changes)
- specifying the args as a plain string
